Automattic/wp-calypso
# Email setup: ensure hidden password reset field is validated

Related to # 8256-gh-Automattic/dotcom-forge

## Proposed Changes

- Ensure that the `PasswordResetEmailValidator` is applied even when the field is hidden, and will display the relevant field if validation fails
- Run early detection for cases where the new email domain matches the recovery email address, and expose the field when the form loads so the user can enter a valid recovery email address
- Abstract the field logic into a hook so it can be reused across the different implementations  of the field for new mailboxes


## Why are these changes being made?

Currently, no validation is run on hidden fields when setting up a new email account. This means if a user doesn't update their reset password email, and it matches the domain name they're creating the account for, they'll end up with a somewhat cryptic error at checkout, and may not know what to do. This PR updates the validation call to include hidden fields, rather than ignoring them, and sets the password reset email field visible if it fails validation. I performed an audit of the other hidden fields we track, they all look adequately typed and/or have sufficient default values that running validation on them when we weren't before won't be a problem.

After implementing the initial validation change, I really didn't like the UX of loading a mailbox creation form with a hidden error we can detect in advance, but still making the user jump through the hoop of clicking "Purchase" before the error is revealed. I experimented with running the validation on the initial render, but that results a form with an error on a prefilled field, which is terrible UX as it makes no sense for us to pre-populate what we know to be an invalid value.

In the end, I've opted for a new hook (open to feedback on the naming and file location) that runs a pseudovalidation on the domain name and user email. If a matching domain name is found, the field is not hidden, and the initial value is blank.

In effect:
- if the user has a wpcom account email that won't break our validation rules, we use it and hide the field as currently happens on `trunk`
- if the user's wpcom account email does break our validation rule, we render the field with no prefilled value, so it appears as simply another part of the signup form
- the updated validation is still in place on this PR, and can serve as a failsafe if an edge case (or perhaps a future flow not using this new hook) results in the field being hidden with an invalid value, clicking "purchase" will validate and expose the field before leading to checkout.


## Testing Instructions

Apologies for the lengthy instructions, it's largely repetition as we cover multiple flows for setting up mailboxes.

First, let's test the updated validation:

<details>
<summary>Begin by applying this test diff:</summary>

```diff
diff --git a/client/my-sites/email/hooks/use-password-reset-email-field.ts b/client/my-sites/email/hooks/use-password-reset-email-field.ts
index 205bd9fb9d..a1a55e0fb1 100644
--- a/client/my-sites/email/hooks/use-password-reset-email-field.ts
+++ b/client/my-sites/email/hooks/use-password-reset-email-field.ts
@@ -17,7 +17,7 @@ export function usePasswordResetEmailField( {
 	selectedDomainName,
 	defaultHiddenFields = [],
 }: PasswordResetEmailFieldProps ): PasswordResetEmailFieldOptions {
-	const userEmail = useSelector( getCurrentUserEmail );
+	const userEmail = 'test@YOURTESTDOMAIN.com';
 
 	// Check if the email is valid prior to official validation so we can
 	// show the field without triggering a validation error when the page
@@ -25,9 +25,7 @@ export function usePasswordResetEmailField( {
 	const isPasswordResetEmailValid = ! new RegExp( `@${ selectedDomainName }$` ).test( userEmail );
 
 	return {
-		hiddenFields: isPasswordResetEmailValid
-			? [ ...defaultHiddenFields, FIELD_PASSWORD_RESET_EMAIL ]
-			: defaultHiddenFields,
-		passwordResetEmailFieldInitialValue: isPasswordResetEmailValid ? userEmail : '',
+		hiddenFields: [ ...defaultHiddenFields, FIELD_PASSWORD_RESET_EMAIL ],
+		passwordResetEmailFieldInitialValue: userEmail,
 	};
 }

```

</details>

- Update the modified `client/my-sites/email/hooks/use-password-reset-email-field.ts` to hardcode the `userEmail` value to an address that matches your test domain (alternatively, you can go in and actually update your wpcom email account to match your test domain), but that can be a pain)
- On a site with a custom domain but no active email products, visit My Home > Upgrades > Email
- Confirm that when the form renders, the Password Reset Email field is hidden, and the textual tip with a "Chang it" link is shown
- Note: the tip will show your actual account email, not the string you just hardcoded. This is expected, the validation is going to run against your temporarily hardcoded test value.
- Enter values for the visible fields on the form
- Click "Purchase"
- Confirm the Password Reset Email field is displayed, with your mocked wpcom email account shown and the correct validation error displayed telling you a different domain is needed.
- Select the Google Workspace card on this same page and repeat the test.
- Confirm the results are the same: no field initially, and after clicking "Purchase" the field is revealed, pre-populated, and showing a validation error

Reset your local branch, but keep the hardcoded email if you'd like, we're going to test with it again

Now, lets test the updated visibility setting to avoid validation errors 

- If needed, update `client/my-sites/email/hooks/use-password-reset-email-field.ts` to hardcode the `userEmail` value to an address that matches your test domain. This will be similar to the previous diff, except it's just one change to the `userEmail` declaration within the hook. Use any address with a domain that matches your test domain.
- On a site with a custom domain but no active email products, visit My Home > Upgrades > Email.
- Confirm that on the Titan card the Password Reset Email field is visible and empty. Test the field to confirm that expected validation ( required field, make sure it's a valid email, make sure it's not the same domain) are being enforced when you click purchase.
- Test and confirm all of the same points for the Google Workspaces card on this same page

Next, we'll test the form used to set up Titan after the email onboarding flow. We'll need a new domain here.

- Update `client/my-sites/email/hooks/use-password-reset-email-field.ts` again, hardcoding `userEmail` to an address using the domain you're going to test the onboarding flow with
- Start at `/start/email-with-onboarding` and proceed through the onboarding flow, adding Professional Email along the way
- After checkout choose the option to set up your email/inbox on the thank you page
- Confirm that the password reset field renders with no prefilled value. Test to confirm the field still validates as a required field , it recognizes invalid email addresses, and it rejects email on the same domain as your new email account.
- Finish adding the initial email account

Now we can test adding additional mailboxes

- On the domain you just tested with Titan, open My Home > Upgrades > Emails and choose the option to add more mailboxes
- Assuming your userEmail is still hardcoded, confirm that the password reset field renders without a prefilled value. Test for expected validation as we've done in the previous steps
